data/transactions/logic/eval.go

```diff
@@ -592,26 +599,234 @@ func (cx *EvalContext) RunMode() RunMode {
+	StackMethodSelector = NewStackType(avmBytes, static(4), "method")
+	// StackStorageKey represents a bytestring that can be used as a key to some storage (global/local/box)
+	StackStorageKey = NewStackType(avmBytes, bound(0, 64), "key")
+	// StackBoxKey represents a bytestring that can be used as a key to a box
+	StackBoxKey = NewStackType(avmBytes, bound(1, 64), "bkey")
+
+	// StackZeroUint64 is a StackUint64 with a minimum value of 0 and a maximum value of 0
+	StackZeroUint64 = NewStackType(avmUint64, bound(0, 0), "0")
+	// StackZeroBytes is a StackBytes with a minimum length of 0 and a maximum length of 0
+	StackZeroBytes = NewStackType(avmUint64, bound(0, 0), "''")
+
+	// AllStackTypes is a list of all the stack types we recognize
+	// so that we can iterate over them in doc prep
+	AllStackTypes = []StackType{
+		StackUint64,
+		StackBytes,
+		StackAny,
+		StackNone,
+		StackBoolean,
+		StackBytes32,
+		StackBigInt,
+		StackMethodSelector,
+		StackStorageKey,
+		StackBoxKey,
+		StackZeroUint64,
+		StackZeroBytes,
+	}
+)
+
+func bound(min, max uint64) [2]uint64 {
+	return [2]uint64{min, max}
+}
+
+func static(size uint64) [2]uint64 {
+	return bound(size, size)
+}
+
+func union(a, b [2]uint64) [2]uint64 {
+	u := [2]uint64{a[0], a[1]}
+	if b[0] < u[0] {
+		u[0] = b[0]
+	}
+
+	if b[1] > u[1] {
+		u[1] = b[1]
+	}
+	return u
+}
+
+// StackType describes the type of a value on the operand stack
+type StackType struct {
+	Name    string
+	AVMType avmType
+	Bound   [2]uint64
+}
+
+// NewStackType Initializes a new StackType with fields passed
+func NewStackType(at avmType, bounds [2]uint64, stname ...string) StackType {
+	name := at.String()
+	if len(stname) > 0 {
+		name = stname[0]
+	}
+
+	return StackType{Name: name, AVMType: at, Bound: bounds}
+}
+
+func (st StackType) union(b StackType) StackType {
+	// TODO: Can we ever receive one or the other
+	// as None? should that be a panic?
+
+	if st.AVMType != b.AVMType {
+		return StackAny
+	}
+
+	// Same type now, so we can just take the union of the bounds
+	return NewStackType(st.AVMType, union(st.Bound, b.Bound))
+}
+
+func (st StackType) narrowed(bounds [2]uint64) StackType {
+	// It's static, set the name to show
+	// the static value
+	if bounds[0] == bounds[1] {
+		switch st.AVMType {
+		case avmBytes:
+			return NewStackType(st.AVMType, bounds, fmt.Sprintf("[%d]byte", bounds[0]))
+		case avmUint64:
+			return NewStackType(st.AVMType, bounds, fmt.Sprintf("%d", bounds[0]))
+		}
+	}
+	return NewStackType(st.AVMType, bounds)
+}
+
+func (st StackType) widened() StackType {
+	// Take only the avm type
+	switch st.AVMType {
+	case avmBytes:
+		return StackBytes
+	case avmUint64:
+		return StackUint64
+	case avmAny:
+		return StackAny
+	default:
+		panic(fmt.Sprintf("What are you tyring to widen?: %+v", st))
+	}
+}
+
+// AssignableTo indicates whether a value typed st might be
+// usable as other.
+func (st StackType) AssignableTo(other StackType) bool {
```

Comment:
This doesn't need to be exported, does it?

I am tempted to call this "overlaps". Each time I reread it, I think like 773 is a bug, because an `any` might not "fit" in type that known to be an int. Then I recall that we really mean something more like "It might work, I can't prove it doesn't".

Reply:
overlaps is reasonable and describes what we're checking